lib/public/Authentication/IProvideUserSecretBackend.php

```diff
@@ -0,0 +1,41 @@
+<?php
+/**
+ * @copyright Copyright (c) 2021, MichaIng <micha@dietpi.com>
+ *
+ * @author MichaIng <micha@dietpi.com>
+ *
+ * @license AGPL-3.0
+ *
+ * This code is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program. If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+// use OCP namespace for all classes that are considered public.
+// This means that they should be used by apps instead of the internal ownCloud classes
+
+namespace OCP\Authentication;
+
+/**
+ * Interface IProvideUserSecretBackend
+ *
+ * @since 23.0.0
+ */
+interface IProvideUserSecretBackend {
+
+	/**
+	 * Optionally returns a stable per-user secret. This secret is for
+	 * instance used to secure file encryption keys.
+	 * @return string
+	 * @since 23.0.0
+	 */
+	public function getCurrentUserSecret(): string;
```

Comment:
Would making this nullable make sense? Especially as `$password = null;` is the default.
```suggestion
	public function getCurrentUserSecret(): ?string;
```

Reply:
I think the idea was to explicitly not allow a null value here, so that checks can be omitted when the interface is implemented. Backends then need to always provide a string.

Ref: https://github.com/nextcloud/server/pull/27929#discussion_r684146422

But I see this is inconsistent with the description "Optionally returns a stable per-user secret.", which just has not been changed when I copy&pasted it from `IApacheBackend` (where it needed to be optional, of course) to its own interface. Shall I replace with:
```
* Returns a stable per-user secret e.g. used to secure file encryption keys.
```

Reply:
Two comments/issues:

1. I've tried the most recent version of this PR with the `IProvideUserSecretBackend` in conjunction with the associated user_saml plugin PR:
https://github.com/nextcloud/user_saml/pull/537
 .
Since users logging in with this method would be using the `user_saml` backend, `lib/private/legacy/OC_User.php:177` never gets executed. Therefore, the password is never set. Am I missing something?
2. With the code from the original PR, there seems to be a problem. After a few minutes, `heartbeat` requests start returning 401 responses. Then, any `getstoragestats` request triggers a page refresh after a 401 response because the session is no longer valid. I have yet to determine what is causing the session to fail, but it is a problem.

Any thoughts? I have a local Keycloak -> NC dev setup working, and I'm happy to test.

Reply:
> Without a password, `getCurrentUserSecret()` will return null. So, this suggested change should either be reversed (probably the recommended solution) or a try/catch should be added

Is it possible to "provide" the `IProvideUserSecretBackend` only when a user secret is actually set? In theory this would then reduce some overhead, when the backend is used without it and would make allowing/checking for `null` obsolete, right? But I clearly lack knowledge about theses things, so probably it's nonsense 😄.

Reply:
I have a proposal to get us unstuck here. How about we make `IProvideUserSecretBackend::getCurrentUserSecret()` nullable, but then do the following:

```
$password = null;
if ($backend instanceof \OCP\Authentication\IProvideUserSecretBackend) {
  $password = $backend->getCurrentUserSecret();
}
if ($password !== null) {
  $userSession->createSessionToken($request, $uid, $uid, $password);
} else {
  $userSession->createSessionToken($request, $uid, $uid);
}
```

This would allow us to disentangle the two problems, preserve the identical behavior for the non-userSecret case and let `createSessionToken` decide what to do when there is no password. This solution would also be more robust, as it removes that dependency on the decision if token passwords are nullable or not and allows us to define an interface that is independent of that.

Reply:
Isn't it exactly the same outcome whether passing `$password = null` or no argument at all? The `createSessionToken` has it's fourth argument `null` in both cases, isn't it?

The main question is indeed whether we want to be more explicit and force usage of a password when the backend is implemented or whether we want to allow implementing the backend without the need to provide a password. Generally I'm fans of being more explicit, but I understand that it may make the backend implementation more complicated.

I think it is agreed that we can or even should merge this PR first, before doing so at the backend, especially since the interfaces are defined here, so the above is really the last open question. I think we need someone who is more in charge to simply do a decision and we'll be able to deal with the result. But we need to get unstuck :smilley:.